virtio-devices/src/net.rs

```diff
@@ -462,25 +465,32 @@ impl Net {
                     true,
                 )
             } else {
-                let mut avail_features = 1 << VIRTIO_NET_F_CSUM
-                    | 1 << VIRTIO_NET_F_CTRL_GUEST_OFFLOADS
-                    | 1 << VIRTIO_NET_F_GUEST_CSUM
-                    | 1 << VIRTIO_NET_F_GUEST_ECN
-                    | 1 << VIRTIO_NET_F_GUEST_TSO4
-                    | 1 << VIRTIO_NET_F_GUEST_TSO6
-                    | 1 << VIRTIO_NET_F_GUEST_UFO
-                    | 1 << VIRTIO_NET_F_HOST_ECN
-                    | 1 << VIRTIO_NET_F_HOST_TSO4
-                    | 1 << VIRTIO_NET_F_HOST_TSO6
-                    | 1 << VIRTIO_NET_F_HOST_UFO
-                    | 1 << VIRTIO_NET_F_MTU
-                    | 1 << VIRTIO_RING_F_EVENT_IDX
-                    | 1 << VIRTIO_F_VERSION_1;
+                let mut avail_features =
+                    1 << VIRTIO_NET_F_MTU | 1 << VIRTIO_RING_F_EVENT_IDX | 1 << VIRTIO_F_VERSION_1;
 
                 if iommu {
                     avail_features |= 1u64 << VIRTIO_F_IOMMU_PLATFORM;
                 }
 
+                if offload_csum {
+                    avail_features |= 1 << VIRTIO_NET_F_CSUM
+                        | 1 << VIRTIO_NET_F_GUEST_CSUM
+                        | 1 << VIRTIO_NET_F_CTRL_GUEST_OFFLOADS;
+
+                    if offload_tso {
+                        avail_features |= 1 << VIRTIO_NET_F_HOST_ECN
+                            | 1 << VIRTIO_NET_F_HOST_TSO4
+                            | 1 << VIRTIO_NET_F_HOST_TSO6
+                            | 1 << VIRTIO_NET_F_GUEST_ECN
+                            | 1 << VIRTIO_NET_F_GUEST_TSO4
+                            | 1 << VIRTIO_NET_F_GUEST_TSO6;
+                    }
+
+                    if offload_ufo {
+                        avail_features |= 1 << VIRTIO_NET_F_HOST_UFO | 1 << VIRTIO_NET_F_GUEST_UFO;
+                    }
+                }
```

Comment:
Due to these interdependencies you need to make `NetConfig::validate()` check that the user isn't providing an invalid configuration.

Reply:
Thanks, fixed and test case added.